File: src/algorithms/graph/connectivity/connected-components/sources/connected-components.go
// Connected Components — find all connected components in an undirected graph using BFS
package connectedcomponents

func connectedComponents(adjacencyList map[string][]string, nodeIds []string) [][]string {
	components := make([][]string, 0) // @step:initialize
	visitedSet := make(map[string]bool) // @step:initialize

	for _, startNodeId := range nodeIds {
		if visitedSet[startNodeId] {
			continue // @step:initialize
		}

		currentComponent := make([]string, 0) // @step:enqueue
		nodeQueue := []string{startNodeId}      // @step:enqueue
		visitedSet[startNodeId] = true          // @step:enqueue

		for len(nodeQueue) > 0 {
			currentNodeId := nodeQueue[0] // @step:dequeue
			nodeQueue = nodeQueue[1:]     // @step:dequeue
			currentComponent = append(currentComponent, currentNodeId) // @step:dequeue,visit

			neighbors := adjacencyList[currentNodeId]
			for _, neighborId := range neighbors {
				if !visitedSet[neighborId] {
					visitedSet[neighborId] = true              // @step:visit-edge
					nodeQueue = append(nodeQueue, neighborId) // @step:visit-edge,enqueue
				}
			}
		}

		components = append(components, currentComponent) // @step:assign-component
	}

	return components // @step:complete
}

File: src/algorithms/graph/connectivity/connected-components/sources/connected-components.ts
// Connected Components — find all connected components in an undirected graph using BFS
function connectedComponents(
  adjacencyList: Record<string, string[]>,
  nodeIds: string[],
): string[][] {
  const components: string[][] = []; // @step:initialize
  const visitedSet = new Set<string>(); // @step:initialize

  for (const startNodeId of nodeIds) {
    if (visitedSet.has(startNodeId)) continue; // @step:initialize

    const currentComponent: string[] = []; // @step:enqueue
    const nodeQueue: string[] = [startNodeId]; // @step:enqueue
    visitedSet.add(startNodeId); // @step:enqueue

    while (nodeQueue.length > 0) {
      const currentNodeId = nodeQueue.shift()!; // @step:dequeue
      currentComponent.push(currentNodeId); // @step:dequeue,visit

      const neighbors = adjacencyList[currentNodeId] ?? [];
      for (const neighborId of neighbors) {
        if (!visitedSet.has(neighborId)) {
          visitedSet.add(neighborId); // @step:visit-edge
          nodeQueue.push(neighborId); // @step:visit-edge,enqueue
        }
      }
    }

    components.push(currentComponent); // @step:assign-component
  }

  return components; // @step:complete
}
